MyHeartCounts/Modules/StudyBundleLoader.swift

```diff
@@ -0,0 +1,178 @@
+//
+// This source file is part of the My Heart Counts iOS application based on the Stanford Spezi Template Application project
+//
+// SPDX-FileCopyrightText: 2025 Stanford University
+//
+// SPDX-License-Identifier: MIT
+//
+
+import Foundation
+import OSLog
+import Spezi
+import SpeziStudyDefinition
+
+@Observable
+final class StudyBundleLoader: Module, Sendable {
+    enum LoadError: Error {
+        case unableToFetchFromServer(any Error)
+        case unableToDecode(any Error)
+        case noLastUsedFirebaseConfig
+    }
+    
+    static let shared = StudyBundleLoader()
+    
+    private let logger = Logger(subsystem: "edu.stanford.MHC.studyLoader", category: "")
+    
+    
+    // Note: we use the Result in here, and set the Task's Failure type to Never, since Task currently only supports type-erased `any Error` failures.
+    // (and does not support typed throws in its init, for whatever reason...)
+    @ObservationIgnored @MainActor private var loadStudyBundleTask: Task<Result<StudyBundle, LoadError>, Never>?
+    // SAFETY: this is only mutated from the MainActor.
+    // NOTE: the compiler thinks the nonisolated(unsafe) isn't needed here. this is a lie. see also https://github.com/swiftlang/swift/issues/81962
+    nonisolated(unsafe) private(set) var studyBundle: Result<StudyBundle, LoadError>?
+//    private var isLoadingStudyBundle = false
+    
+    // SAFETY: the FileManager itself is not thread safe, but we have our own instance (as opposed to using `FileManager.default`),
+    // and we never mutate it.
+    nonisolated(unsafe) private let fileManager = FileManager()
+    
+    private let studyBundlesUrl: URL
+    
+//    // SAFETY: this is only mutated from the MainActor.
+//    // NOTE: the compiler thinks the nonisolated(unsafe) isn't needed here. this is a lie. see also https://github.com/swiftlang/swift/issues/81962
+//    nonisolated(unsafe) private(set) var consentDocument: Result<String, LoadError>?
+    
+    private init() {
+        studyBundlesUrl = URL.documentsDirectory.appending(path: "MHC/StudyBundlesCache", directoryHint: .isDirectory)
+        Task {
+            _ = try? await update()
+        }
+    }
+    
+    
+    @discardableResult
+    @MainActor
+    func update() async throws(LoadError) -> StudyBundle {
```

Comment:
a downside of the new (StudyBundle-based) approach we now use is that we need to download the whole bundle in order to determine if the on-server one is newer than what we currently have.

this was the same with the old JSON-file-based approach, and it's not that bad for the time being, since the StudyBundle currently is rather small (~22KB), but if we choose to include additional files in the bundle in the future (eg: images, videos, etc), this might become a problem. it might make sense to encode the study revision into the file metadata in Firebase, but you can't set that when uploading via the web interface

Reply:
I think this is fine for now; in the future and for a study app, we might download the bundles from a server that should have an endpoint that lists all study bundles and their versions that the user has access to get their download URLs and that endpoint could then also expose the version. I think we can live with that overhead for now and in MHC.